### PR TITLE
mod(es): change definition of optional chain

### LINKED ISSUE #3610: Optional get private field wrongly transformed
### Describe the bug

Somehow it's transformed into `_classPrivateFieldGet === null`

### Input code

```typescript
class MyClass {
  #a
  foo(o) {
    o?.#a
  }
}
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "ecmascript",
      "jsx": true
    },
    "transform": {
      "legacyDecorator": true,
      "decoratorMetadata": false
    },
    "target": "es2016",
    "loose": false,
    "minify": {
      "compress": false,
      "mangle": false
    }
  },
  "module": {
    "type": "es6"
  },
  "minify": false,
  "isModule": true,
  "env": {
    "targets": "chrome 70"
  }
}
```

### Playground link

https://play.swc.rs/?version=1.2.141&code=H4sIAAAAAAAAA0vOSSwuVvCtdAbT1VwKCsqJQCItP18jXxPMV1DIt9cDC9Zy1QIAreyLqS8AAAA%3D&config=H4sIAAAAAAAAA0WOuw6DMAxFf6XyzEA7UIm5Kx9hBUOpyEO2qRqh%2FHsTHmKzrs899gofMdCuEJCFuEwSneIPWiBjUQxPQaHKWI6UF0oVKKOTwbMt%2BEwjmvgi4xnV8w5V0J9BR4o9KkI74CxbHXkkLQfkUd%2BbLJ%2B9FzqACuzkpiEWt%2FE2MIlcK3TjfJIpu6zvlxKsoDHQ5mwgXY6jN0l3gPt35L5bZ%2Fsk68G82Vu6PWtI6Q%2BymPZHFAEAAA%3D%3D

### Expected behavior

Same as babel do

### Actual behavior

_No response_

### Version

1.2.141

### Additional context

_No response_

### LINKED ISSUE #3644: Wrongly transform private field access in optional call
### Describe the bug

Would transpile into
```js
(ref = (_obj = _classPrivateFieldGet(this, _fieldFunc)).call) === null || ref === void 0 ? void 0 : ref.call(_obj, this);
```
which access `_classPrivateFieldGet(this, _fieldFunc)).call` and throws

### Input code

```typescript
class A {
    #fieldFunc;
    x = 1;
    test() {
        this.#fieldFunc?.();
    }
}
```

### Config

_No response_

### Playground link

https://play.swc.rs/?version=1.2.133&code=H4sIAAAAAAAAA0vOSSwuVnBUqOZSAALltMzUnBS30rxkazC%2FQsFWwRDCLEktLtHQhKoDC2RkFushNNjraWhCVNZy1QIAaypvMVYAAAA%3D&config=H4sIAAAAAAAAA0WMSwrDMAxE76K1F00WpeQOPYRwleDiH5ICNcZ3r11cshtm3rwKb7GwVcjIQjySlKj4gQ3IBhTLLiuYjvVqRy%2FUDCjyQToQWW%2FLo88%2BJaEJGAguur0MmU0hM4lcE8bD%2F8nWXSG9zlFU0JLp57xDuxzz5%2BQ5QeWT2he%2FL5pQuAAAAA%3D%3D

### Expected behavior

check `_classPrivateFieldGet(this, _fieldFunc))` directly instead of `.call`

### Actual behavior

_No response_

### Version

1.2.143

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Austaras**
**Description:**

Per spec, the base of optional chain can only be either:
1. an member expression
2. an call expression which callee could only be an normal expression, not an super or import

**BREAKING CHANGE:**

AST definition changed

**Related issue (if exists):**
Would closes #3644 and closes #3610

### CODE REVIEW

File: crates/swc_ecma_transforms_compat/src/es2015/regenerator/case.rs [hunk 1]
```diff
@@ -338,8 +338,23 @@ impl CaseHandler<'_> {
             | Expr::PrivateName(..)
             | Expr::Invalid(..) => e,
 
+            // opt chain should already be transformed at this point but anyway
```
Comment (Austaras) [inline crates/swc_ecma_transforms_compat/src/es2015/regenerator/case.rs:341]: Can we simply drop optional call here as it's es2020 feature?

Comment (kdy1) [reply crates/swc_ecma_transforms_compat/src/es2015/regenerator/case.rs:342]: I think so, will you do it?

Comment (Austaras) [reply crates/swc_ecma_transforms_compat/src/es2015/regenerator/case.rs:342]: Done

File: crates/swc_ecma_ast/src/expr.rs [hunk 2]
```diff
@@ -1013,7 +1013,87 @@ impl Take for PatOrExpr {
...
+}
+
+#[ast_node]
+#[derive(Eq, Hash, Is, EqIgnoreSpan)]
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+pub enum OptChainBase {
+    #[tag("MemberExpression")]
+    Member(MemberExpr),
+    #[tag("CallExpression")]
+    Call(OptCall),
```
Comment (kdy1) [inline crates/swc_ecma_ast/src/expr.rs:1026]: Can't we use `CallExpr` instead?

Comment (kdy1) [reply crates/swc_ecma_ast/src/expr.rs:1026]: Nevermind, the type of `callee` is different.

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you so much!

**Review by kdy1 — APPROVED** [review]
Thank you so much for your hard work!

---

swc-bump:
 - swc_ecma_ast --breaking

**Pull request merged**